src/libraries/System.Private.CoreLib/src/System/Threading/ThreadInt64PersistentCounter.cs

```diff
@@ -2,79 +2,69 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections.Generic;
 using System.Diagnostics;
 using System.Runtime.CompilerServices;
+using Internal.Runtime.CompilerServices;
 
 namespace System.Threading
 {
     internal sealed class ThreadInt64PersistentCounter
     {
-        // This type is used by Monitor for lock contention counting, so can't use an object for a lock. Also it's preferable
-        // (though currently not required) to disallow/ignore thread interrupt for uses of this lock here. Using Lock directly
-        // is a possibility but maybe less compatible with other runtimes. Lock cases are relatively rare, static instance
-        // should be ok.
         private static readonly LowLevelLock s_lock = new LowLevelLock();
 
-        private readonly ThreadLocal<ThreadLocalNode> _threadLocalNode = new ThreadLocal<ThreadLocalNode>(trackAllValues: true);
         private long _overflowCount;
 
+        // This should be a HashSet, that type is currently not available in CoreLib
+        private Dictionary<ThreadLocalNode, bool> _nodes = new Dictionary<ThreadLocalNode, bool>();
+
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public void Increment()
+        public static void Increment(object threadLocalCountObject)
         {
-            ThreadLocalNode? node = _threadLocalNode.Value;
-            if (node != null)
-            {
-                node.Increment();
-                return;
-            }
-
-            TryCreateNode();
+            Debug.Assert(threadLocalCountObject is ThreadLocalNode);
+            Unsafe.As<ThreadLocalNode>(threadLocalCountObject).Increment();
         }
 
-        [MethodImpl(MethodImplOptions.NoInlining)]
-        private void TryCreateNode()
+        public object CreateThreadLocalCountObject()
         {
-            Debug.Assert(_threadLocalNode.Value == null);
+            var node = new ThreadLocalNode(this);
 
+            s_lock.Acquire();
```

Comment:
any reason not to acquire the lock within try/finally?

Reply:
Since thread-abort is not supported it doesn't make much difference. Only thing is if Acquire were to throw (this particular method currently doesn't allocate) then the call to Release in the finally block would be invalid. In order to weed out those issues when calling it inside the try block it would have to return something or have an `out acquired` parameter (similar solution for handling thread-abort). It feels simpler to call it outside the try, since Acquire would already have a contract where if it throws, then it must not have acquired the lock, though even that may be difficult to guarantee for all types of exceptions.

Reply:
Thread abort is actually supported internally - debugger funceval support uses it to abort long running or hung methods or properties.

Reply:
We do not make guarantees about the state of the process when debugger thread abort kicks in.

Reply:
Based on https://github.com/dotnet/coreclr/pull/8551 and https://github.com/dotnet/coreclr/pull/8949 it looks like the JIT would remove try/finally blocks where the try block is empty and the finally block has code. So I figure there would not be a way to harden for thread-abort during func-eval, unless there is some other way.